src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -158,28 +154,34 @@ public void insertRecordToBuffer(SinkRecord record) {
     }
   }
 
-  /* For EOS, fetch the offset from snowflake during only during first time after connector starts and record is received in Kafka Connector */
-  @VisibleForTesting
-  protected void fetchLastCommittedOffsetToken() {
-    String offsetToken = this.channel.getLatestCommittedOffsetToken();
-    try {
-      if (offsetToken == null) {
-        this.offsetPersistedInSnowflake = NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE;
-      } else {
-        this.offsetPersistedInSnowflake = Long.parseLong(offsetToken);
-      }
+  /**
+   * Pre-computes the offset token which might have been persisted into Snowflake for this channel.
+   *
+   * <p>It is possible the offset token is null. In this case, -1 is set
+   *
+   * <p>Note: This code is execute only for the first time connector starts or restarts after task
+   * was killed previously.
+   *
+   * <p>This doesnt get invoked on rebalance.
+   *
+   * @param record record this partition received.
+   */
+  private void initChannel(final SinkRecord record) {
+    if (!hasChannelReceivedAnyRecordsBefore) {
```

Comment:
don't we always recreate the TopicPartitionChannel object during start/rebalance? If yes, this should always be false?

Reply:
Correct but we call this from insertRecord. So from insertRecord, `hasChannelReceivedAnyRecordsBefore` will be false only once. (First record)

Reply:
FYI, I changed this name to `precomputeOffsetTokenForChannel`